File: 284.peeking_iterator/iterator.go
// 284. Peeking Iterator
// Topics: 'Array', 'Design', 'Iterator'
// Level: 'Medium'

// Design an iterator that supports the peek operation on an existing iterator in addition to the hasNext and the next operations.

// Implement the PeekingIterator class:

//     PeekingIterator(Iterator<int> nums) Initializes the object with the given integer iterator iterator.
//     int next() Returns the next element in the array and moves the pointer to the next element.
//     boolean hasNext() Returns true if there are still elements in the array.
//     int peek() Returns the next element in the array without moving the pointer.

// Note: Each language may have a different implementation of the constructor and Iterator, but they all support the int next() and boolean hasNext() functions.

// Example 1:

// Input
// ["PeekingIterator", "next", "peek", "next", "next", "hasNext"]
// [[[1, 2, 3]], [], [], [], [], []]
// Output
// [null, 1, 2, 2, 3, false]

// Explanation
// PeekingIterator peekingIterator = new PeekingIterator([1, 2, 3]); // [1,2,3]
// peekingIterator.next();    // return 1, the pointer moves to the next element [1,2,3].
// peekingIterator.peek();    // return 2, the pointer does not move [1,2,3].
// peekingIterator.next();    // return 2, the pointer moves to the next element [1,2,3]
// peekingIterator.next();    // return 3, the pointer moves to the next element [1,2,3]
// peekingIterator.hasNext(); // return False

// Constraints:

//     1 <= nums.length <= 1000
//     1 <= nums[i] <= 1000
//     All the calls to next and peek are valid.
//     At most 1000 calls will be made to next, hasNext, and peek.

package peekingiterator

type Iterator[T any] struct {
	arr []T
	i   int
}

func (i *Iterator[T]) hasNext() bool {
	return i.i < len(i.arr)
}

func (i *Iterator[T]) next() T {
	v := i.arr[i.i]
	i.i++
	return v
}

type PeekingIterator[T any] struct {
	arr  []T
	iter *Iterator[T]
	i    int
}

func Constructor[T any](iter *Iterator[T]) *PeekingIterator[T] {
	return &PeekingIterator[T]{
		i:    0,
		arr:  []T{},
		iter: iter,
	}
}

func (p *PeekingIterator[T]) hasNext() bool {
	return p.iter.hasNext() || p.i < len(p.arr)
}

func (p *PeekingIterator[T]) next() T {
	if p.i < len(p.arr) {
		v := p.arr[p.i]
		p.i++
		return v
	}
	return p.iter.next()
}

func (p *PeekingIterator[T]) peek() T {
	if p.i < len(p.arr) {
		return p.arr[p.i]
	}
	v := p.iter.next()
	p.arr = append(p.arr, v)
	return v
}

File: 341.flatten_nested_list_iterator/iterator.py
# 341. Flatter Nested List Iterator
# Topics: 'Stack', 'Tree', 'Depth-First Search', 'Design', 'Queue', 'Iterator'
# Level: 'Medium'

# You are given a nested list of integers nestedList. Each element is either an integer or a list whose elements may also be integers or other lists. Implement an iterator to flatten it.

# Implement the NestedIterator class:

#     NestedIterator(List<NestedInteger> nestedList) Initializes the iterator with the nested list nestedList.
#     int next() Returns the next integer in the nested list.
#     boolean hasNext() Returns true if there are still some integers in the nested list and false otherwise.

# Your code will be tested with the following pseudocode:

# initialize iterator with nestedList
# res = []
# while iterator.hasNext()
#     append iterator.next() to the end of res
# return res

# If res matches the expected flattened list, then your code will be judged as correct.

 

# Example 1:

# Input: nestedList = [[1,1],2,[1,1]]
# Output: [1,1,2,1,1]
# Explanation: By calling next repeatedly until hasNext returns false, the order of elements returned by next should be: [1,1,2,1,1].

# Example 2:

# Input: nestedList = [1,[4,[6]]]
# Output: [1,4,6]
# Explanation: By calling next repeatedly until hasNext returns false, the order of elements returned by next should be: [1,4,6].

 

# Constraints:

#     1 <= nestedList.length <= 500
#     The values of the integers in the nested list is in the range [-106, 106].

# """
# This is the interface that allows for creating nested lists.
# You should not implement it, or speculate about its implementation
# """
# class NestedInteger:
#    def isInteger(self) -> bool:
#        """
#        @return True if this NestedInteger holds a single integer, rather than a nested list.
#        """

#    def getInteger(self) -> int:
#        """
#        @return the single integer that this NestedInteger holds, if it holds a single integer
#        Return None if this NestedInteger holds a nested list
#        """

#    def getList(self) -> [NestedInteger]:
#        """
#        @return the nested list that this NestedInteger holds, if it holds a nested list
#        Return None if this NestedInteger holds a single integer
# #        """

# crazy good solution
# extend stack if not a number
# lazy initialization
class NestedIterator:
    def __init__(self, nestedList: [NestedInteger]):
        self.stack = nestedList[::-1]

    def next(self) -> int:
        return self.stack.pop().getInteger()
            
    def hasNext(self) -> bool:
        while self.stack:
            top = self.stack[-1]
            if top.isInteger():
                return True
            self.stack.extend(self.stack.pop().getList()[::-1])
        return False


# Lazy initialization each NestedInteger tree
# class NestedIterator:
#     def __init__(self, nestedList: [NestedInteger]):
#         self.i = 0
#         self.nestedList = nestedList
#         self.q = deque()

#     def next(self) -> int:
#         return self.q.popleft()
            
#     def hasNext(self) -> bool:
#         while not self.q:
#             if self.i >= len(self.nestedList):
#                 break
#             self.dfs(self.nestedList[self.i])
#             self.i+=1
#         return self.q

#     def dfs(self, root: NestedInteger):
#         if not root:
#             return
#         else:
#             if root.isInteger():
#                 self.q.append(root.getInteger())
#             else:
#                 for ni in root.getList():
#                     self.dfs(ni)

# Eager initialization
# class NestedIterator:
#     def __init__(self, nestedList: [NestedInteger]):
#         self.arr = []
#         self.i = 0
#         def dfs(root: NestedInteger):
#             if not root:
#                 return
#             if root.isInteger():
#                 self.arr.append(root.getInteger())
#             else:
#                 for ni in root.getList():
#                     dfs(ni)
#         for ni in nestedList:
#             dfs(ni)
            

#     def next(self) -> int:
#         v = self.arr[self.i]
#         self.i+=1
#         return v
    
#     def hasNext(self) -> bool:
#         return self.i < len(self.arr)
         
